src/mpid/ch4/shm/posix/eager/iqueue/iqueue_types.h

```diff
@@ -0,0 +1,97 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef POSIX_EAGER_IQUEUE_TYPES_H_INCLUDED
+#define POSIX_EAGER_IQUEUE_TYPES_H_INCLUDED
+
+#include <mpidimpl.h>
+#include "mpidu_shm.h"
+
+#define MPIDI_POSIX_EAGER_IQUEUE_DEFAULT_CELL_SIZE (64 * 1024)
+
+#define MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL 0
+#define MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD 1
+#define MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL 2
+
+typedef struct MPIDI_POSIX_EAGER_IQUEUE_cell MPIDI_POSIX_EAGER_IQUEUE_cell_t;
+
+/* Each cell contains some data being communicated from one process to another. */
+struct MPIDI_POSIX_EAGER_IQUEUE_cell {
+    uint16_t type;              /* Type of cell (head/tail/etc.) */
+    uint16_t from;              /* Who is the message in the cell from */
+    uint32_t payload_size;      /* Size of the message in the cell */
+    uintptr_t prev;             /* Internal pointer when finding the previous active cell */
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *next;      /* Internal pointer when finding the next active cell */
+    MPIDI_POSIX_am_header_t am_header;  /* If this cell is the beginning of a message, it will have
```

Comment:
Why `prev` is `uintptr_t ` but `next` is `MPIDI_POSIX_EAGER_IQUEUE_cell_t ` ?

Reply:
Prev is an index in the array while next is a pointer to the cell itself. I don't know why that is right now. I'll have to go try to figure it out again.

Reply:
The "prev" is the uintptr_t because it is the absolute offset of a cell in the shared memory segment. Offset is absolute, the same over all ranks. The field "prev" architecturally has a public scope of visibility shared between all ranks.

The "next" is a usual pointer because it is a private process memory virtual address. Different process have different virtual addresses of the same shared memory cell. The field "next" architecturally has a private scope of visibility inside the rank which in the current moment process the cell.

Reply:
I see. Could you please add short description in code comment ?

Reply:
Done.